### ISSUE TITLE
'Operation not permitted' when trying to connect with /var/cfm_socket

### ISSUE THREAD

**Issue by kr1tzy** [qilingframework/qiling#559]
I've been using the [TendaAC1518](https://github.com/qilingframework/qiling/blob/master/examples/tendaac1518_httpd.py) example as a reference for another Tenda device and am continuously getting Operation not permitted when httpd tries to connect to /var/cfm_socket. 

Some things I've tried:
- hook the connect calls to verify the correct path for /var/cfm_socket in libCfm.so 
- create the cfm_socket on my host machine and map a new filesystem entry with Qiling
- running it as root, debugging it, toggling various ql flags, etc.
- implement fork and ipc syscall (both are called and not implemented by default)
- overwrite port 80 with 8080 (i know Qiling should do this by default but it wasn't working in my case)
- patch the libChipApi.so and the httpd binary with a valid NIC from my host
- change default nvram values in the config files

Other details:
- Host: x86_64 Ubuntu 18.04 VM
- Device/firmware: Tenda FH1202 / v1.2.0.20 
- Arch: mipsel-32bit

Standard console output
![image](https://user-images.githubusercontent.com/14809526/96935796-ceaa9a80-1492-11eb-8c6a-a680afdea11b.png)

Output of my connect hook
![image](https://user-images.githubusercontent.com/14809526/96945650-5224b580-14ac-11eb-9299-579797c5a913.png)

Code
```python
#!/usr/bin/python3

import os, sys, socket, threading
from qiling import *
from qiling.os.posix import syscall
import signal

def patcher(ql):
    br0_addr = ql.mem.search("br0".encode() + b"\x00")
    for addr in br0_addr:
        ql.mem.write(addr, b'ens33\x00')


def nvram_listener():
    server_address = 'squashfs-root/var/cfm_socket'
    data = ""

    try:  
        os.unlink(server_address)  
    except OSError:  
        if os.path.exists(server_address):  
            raise  

    # Create UDS socket  
    sock = socket.socket(socket.AF_UNIX,socket.SOCK_STREAM)  
    sock.bind(server_address)  
    sock.listen(1)  

    while True:  
        connection, client_address = sock.accept()  
        try:  
            while True:  
                data += str(connection.recv(1024))

                if "lan.webiplansslen" in data:  
                    connection.send('192.168.170.169'.encode())  
                else: 
                    break  
                data = ""
        finally:  
                connection.close() 

def my_ipc(ql, *rest):
    pass

def my_fork(ql):
    regreturn = 0
    ql.nprint("vfork() = %d" % regreturn)
    ql.os.definesyscall_return(regreturn)

def hook_socketListen(ql, *rest):
    print("-" * 64)
    print("Socket listen: " + str(rest))
    print(f"Port: {ql.reg.a1}")
    ql.reg.a1 = 8080

    for i in range(0, 25):
        if ql.stack_read(i*4) == 80:
            ql.stack_write(i*4, 8080)
            print(f"{ql.stack_read(i*4)}")
    print("-" * 64)

def my_connect(ql, *rest):
    print(f"[*] connect: {str(rest)}")
    print(f"\tfd: {rest[0]}")
    print(f"\tsockaddr*: {ql.mem.read(rest[1], 100)}")
    print(f"\taddress length: {rest[2]}")


def tenda_fh1220(path, rootfs):
    ql = Qiling(path, rootfs, output="debug", log_dir="qlog", console=True) 
    ql.root             = True
    ql.bindtolocalhost  = True
    ql.debug_stop       = False
    # ql.debugger         = True
    ql.add_fs_mapper('/proc', '/proc')
    ql.add_fs_mapper("/dev/urandom","/dev/urandom")
    ql.hook_address(patcher, ql.loader.elf_entry)
    ql.patch(0x00004de8, b"ens33\x00", file_name = b'libChipApi.so')
    # ql.hook_address(hook_socketListen, 0x41a518)
    # ql.set_syscall(4002, my_fork)
    # ql.set_syscall(4117, my_ipc)
    ql.set_syscall(4170, my_connect)
    ql.run()


if __name__ == "__main__":
    # nvram_listener_thread = threading.Thread(target=nvram_listener, daemon=True)
    # nvram_listener_thread.start()
    tenda_fh1220(["squashfs-root/bin/httpd", "--debugger", "--verbose", "--home", "webroot"], "squashfs-root")
```

**Comment by wtdcode** [comment]
Note that you may map cfm_socket to your custom object as the second parameter of add_fs_mapper.

**Comment by kr1tzy** [comment]
> Note that you may map cfm_socket to your custom object as the second parameter of add_fs_mapper.

I tried that (second bullet point under things i tried). I just removed it before posting the code on here. It looked like this:

```python
ql.add_fs_mapper("/var/cfm_socket","/home/kritz/Playground/firmware/tenda/_US_FH1202V1.0BR_V1.2.0.20(408)_EN_TD.bin.extracted/cfm_socket")
```

**Comment by kr1tzy** [comment]
Ended up tracing this into the `bind()` of the 'cfmd' binary and found it was returning an 'AF_UNIX path too long' error. It seems this is specific to Python and its AF_UNIX socket implementation because a lot of other projects are experiencing the same issue (https://www.google.com/search?q=AF_UNIX+path+too+long). To fix it I renamed the `_US_FH1202V1.0BR_V1.2.0.20(408)_EN_TD.bin.extracted` directory to `fh1202` and it started working. I appreciate the quick reply to help out @wtdcode!
